core/src/main/java/com/linecorp/armeria/internal/annotation/AnnotatedBeanFactoryRegistry.java

```diff
@@ -331,4 +324,65 @@ public String toString() {
                               .toString();
         }
     }
+
+    static final class AnnotatedBeanFactory<T> {
+
+        private final BeanFactoryId beanFactoryId;
+        private final Entry<Constructor<T>, List<AnnotatedValueResolver>> constructor;
+        private final Map<Field, AnnotatedValueResolver> fields;
+        private final Map<Method, List<AnnotatedValueResolver>> methods;
+
+        private AnnotatedBeanFactory(BeanFactoryId beanFactoryId,
+                                     Entry<Constructor<T>, List<AnnotatedValueResolver>> constructor,
+                                     Map<Method, List<AnnotatedValueResolver>> methods,
+                                     Map<Field, AnnotatedValueResolver> fields) {
+            this.beanFactoryId = requireNonNull(beanFactoryId, "beanFactoryId");
+            this.constructor = simpleImmutableEntry(requireNonNull(constructor, "constructor"));
+            this.fields = ImmutableMap.copyOf(requireNonNull(fields, "fields"));
+            this.methods = ImmutableMap.copyOf(requireNonNull(methods, "methods"));
+        }
+
+        T create(ResolverContext resolverContext) {
+            try {
+                final Object[] constructorArgs = AnnotatedValueResolver.toArguments(
+                        constructor.getValue(), resolverContext);
+                final T instance = constructor.getKey().newInstance(constructorArgs);
+
+                for (final Entry<Method, List<AnnotatedValueResolver>> method : methods.entrySet()) {
+                    final Object[] methodArgs = AnnotatedValueResolver.toArguments(
+                            method.getValue(), resolverContext);
+                    method.getKey().invoke(instance, methodArgs);
+                }
+
+                for (final Entry<Field, AnnotatedValueResolver> field : fields.entrySet()) {
+                    final Object fieldArg = field.getValue().resolve(resolverContext);
+                    field.getKey().set(instance, fieldArg);
+                }
+
+                return instance;
+            } catch (Throwable cause) {
+                throw new IllegalArgumentException(
+                        "cannot instantiate a new object: " + beanFactoryId, cause);
+            }
+        }
+
+        private static <K, V> SimpleImmutableEntry<K, V> simpleImmutableEntry(Entry<K, V> entry) {
```

Comment:
How about changing the method signature to: `Entry<K, V> immutableEntry(Entry<K, V>)`

Reply:
How about removing this method and inlining into constructor?

Reply:
That was in the constructor, but I just wanted to leave it out.
Let me do what @trustin says. :)